src/main/java/org/testng/ITestNGMethod.java

```diff
@@ -15,233 +14,252 @@
-
-  /**
-   * @return the success percentage for this method (between 0 and 100).
-   */
-  int getSuccessPercentage();
-
-  /**
-   * @return The id of the thread this method was run in.
-   */
-  String getId();
-
-  void setId(String id);
-
-  long getDate();
-
-  void setDate(long date);
-
-  /**
-   * Returns if this ITestNGMethod can be invoked from within IClass.
-   */
-  boolean canRunFromClass(IClass testClass);
-
-  /**
-   * @return true if this method is alwaysRun=true
-   */
-  boolean isAlwaysRun();
-
-  /**
-   * @return the number of threads to be used when invoking the method on parallel
-   */
-  int getThreadPoolSize();
-
-  void setThreadPoolSize(int threadPoolSize);
-
-  boolean getEnabled();
-
-  String getDescription();
-  void setDescription(String description);
-
-  void incrementCurrentInvocationCount();
-  int getCurrentInvocationCount();
-  void setParameterInvocationCount(int n);
-  int getParameterInvocationCount();
-  void setMoreInvocationChecker(Callable<Boolean> moreInvocationChecker);
-  boolean hasMoreInvocation();
-
-  ITestNGMethod clone();
-
-  IRetryAnalyzer getRetryAnalyzer();
-  void setRetryAnalyzer(IRetryAnalyzer retryAnalyzer);
-
-  boolean skipFailedInvocations();
-  void setSkipFailedInvocations(boolean skip);
-
-  /**
-   * The time under which all invocationCount methods need to complete by.
-   */
-  long getInvocationTimeOut();
-
-  boolean ignoreMissingDependencies();
-  void setIgnoreMissingDependencies(boolean ignore);
-
-  /**
-   * Which invocation numbers of this method should be used (only applicable
-   * if it uses a data provider). If this value is an empty list, use all the values
-   * returned from the data provider.  These values are read from the XML file in
-   * the <include invocationNumbers="..."> tag.
-   */
-  List<Integer> getInvocationNumbers();
-  void setInvocationNumbers(List<Integer> numbers);
-
-  /**
-   * The list of invocation numbers that failed, which is only applicable for
-   * methods that have a data provider.
-   */
-  void addFailedInvocationNumber(int number);
-  List<Integer> getFailedInvocationNumbers();
-
-  /**
-   * The scheduling priority. Lower priorities get scheduled first.
-   */
-  int getPriority();
-  void setPriority(int priority);
-
-  /**
-   * @return the XmlTest this method belongs to.
-   */
-  XmlTest getXmlTest();
-
-  ConstructorOrMethod getConstructorOrMethod();
-
-  /**
-   * @return the parameters found in the include tag, if any
-   * @param test - The {@link XmlTest} object.
-   */
-  Map<String, String> findMethodParameters(XmlTest test);
-  
-  /**
-   * getRealClass().getName() + "." +  getMethodName()
-   * @return qualified name for this method
-   */
-  String getQualifiedName();
+    /**
+     * @return The real class on which this method was declared
+     * (can be different from getMethod().getDeclaringClass() if
+     * the test method was defined in a superclass).
+     */
+    Class getRealClass() ;

```

Comment:
Class<?>

Reply:
Oh, then better: `Class<? extends ITestClass>`

Reply:
@juherr but there're 2 different types used -> `ITestClass` and `ITestNGMethod`... if we set `Class<? extends ITestClass>`, then it won't be possible to use it with `ITestNGMethod`...
may be we should use generics instead?
```java
<T> Class<T> getRealClass();
...
Class<ITestNGMethod> methodClass = method.getRealClass();
...
Class<ITestClass> cls = m.getRealClass();
```

Reply:
Ok. I think we already had a similar discussion with @krmahadevan. It smells but fixing it is not the goal of this PR. So keep Class.